File: app/src/main/java/com/tryden/leetcode/java/easy/ClimbingStairs.java
package com.tryden.leetcode.java.easy;


/*
 *
 * [70] Climbing Stairs
 *
 * https://leetcode.com/problems/climbing-stairs/description/
 *
 */

public class ClimbingStairs {
    public static void main(String[] args) {
        int n = 11;
        climbStairs(n);
    }
    public static int climbStairs(int n) {
        int a = 1, b = 1;
        while (n-- > 0) {
            a = (b += a) - a;
        }
        return a;
    }
}

File: app/src/main/java/com/tryden/leetcode/kotlin/easy/ClimbingStairs.kt
package com.tryden12.leetcode_kotlin.easy

/*
 *
 * [70] Climbing Stairs
 *
 * https://leetcode.com/problems/climbing-stairs/description/
 *
 */

    fun main() {

    }


    private val cache = mutableMapOf<Int, Int>()

    private fun climbStairs(n: Int) : Int = when {
        n < 1  -> 0
        n == 1 -> 1
        n == 2 -> 2
        else -> cache.getOrPut(n) {
            climbStairs(n-1) + climbStairs(n-2)
        }
    }